cmd/cli/version.go

```diff
@@ -1,32 +1,146 @@
 package main
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
 	"io"
+	"strconv"
 
+	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
+	"k8s.io/client-go/kubernetes"
 
+	"github.com/openservicemesh/osm/pkg/constants"
+	httpserverconstants "github.com/openservicemesh/osm/pkg/httpserver/constants"
 	"github.com/openservicemesh/osm/pkg/version"
 )
 
 const versionHelp = `
-This command prints out all the version information used by OSM
+This command prints the OSM CLI and remote mesh version information
 `
 
-// PrintCliVersion prints the version
-func PrintCliVersion(out io.Writer) {
-	_, _ = fmt.Fprintf(out, "Version: %s; Commit: %s; Date: %s\n", version.Version, version.GitCommit, version.BuildDate)
+type versionCmd struct {
+	out           io.Writer
+	namespace     string
+	remote        bool
+	clientset     kubernetes.Interface
+	remoteVersion remoteVersionGetter
+}
+
+type remoteVersionGetter interface {
+	proxyGetMeshVersion(pod string, namespace string, clientset kubernetes.Interface) (*version.Info, error)
+}
+
+type remoteVersion struct{}
+
+type remoteVersionInfo struct {
+	meshName string
+	version  *version.Info
+}
+
+type versionInfo struct {
+	cliVersionInfo    *version.Info
+	remoteVersionInfo *remoteVersionInfo
 }
 
 func newVersionCmd(out io.Writer) *cobra.Command {
+	versionCmd := &versionCmd{
+		out: out,
+	}
 	cmd := &cobra.Command{
 		Use:   "version",
-		Short: "osm cli version",
+		Short: "osm cli and mesh version",
 		Long:  versionHelp,
 		Args:  cobra.NoArgs,
 		Run: func(cmd *cobra.Command, args []string) {
```

Comment:
Updated `Run` to `RunE`. I had originally kept `Run` here since I wanted to format the error to indicate that the failure occurred when getting the mesh version. There is no reason I can't wrap the error and then return it as an error.

After changes - an example output when getting the mesh version fails:
```shell
CLI Version: version.Info{Version:"dev", GitCommit:"5dcc0d0f087f99205cc85a220568432b234b048c", BuildDate:""}
Error: Failed to get mesh version: No mesh found in namespace [kube-system]
```

Previous output:
```shell
CLI Version: version.Info{Version:"dev", GitCommit:"5dcc0d0f087f99205cc85a220568432b234b048c", BuildDate:""}
Failed to get mesh version: No mesh found in namespace [kube-system]
```

Reply:
The "Error:" is slightly redundant, but returning an error here will also ultimately result in a non-zero exit code returned to the shell. Getting that to happen with plain `Run` wouldn't be as easy.